### ISSUE TITLE
OpenSSL 3 upgrade: [BREAKING CHANGES]

### ISSUE THREAD

**Issue by thunder-coding** [termux/termux-packages#9050]
This is to inform you about the new OpenSSL 3 upgrade on which we are currently working on. This new version of OpenSSL 3 brings a lot of breaking changes which can seriously affect you.

**If you care about stability, consider not upgrading your packages until OpenSSL 3 rebuilds are complete**.

You may also consider switching to Grimler's mirrors using `termux-change-repo` until the rebuild is done as they won't receive updates until the rebuild work is done

The OpenSSL 3 rebuild will start from **Saturday 26 Feb 2022 10:00 AM IST**

# What you need to know as a normal Termux user

- OpenSSL will be upgraded to 3.x breaking any software that isn't packaged by us.
- Packages are expected to be broken for a few days until every package is rebuilt against the newer version of OpenSSL.
- You might also need to reinstall any packages using native libraries installed using `pip`, `npm`, `yarn` and friends.
- Also if you are using any package built by yourself make sure you rebuild it

# Legacy openssl-1.1

When openssl will be upgraded to 3.x, `openssl-1.1` will be added for backward compatibility of applications still relying on OpenSSL 1.1.

## Building packages that still need legacy openssl 1.1

Make sure you have `openssl-1.1` installed:
```sh
pkg install openssl-1.1
```

You'll need to run the configure command for CMake/Make/Meson/etc as:

```
CFLAGS=" -I$PREFIX/include/openssl-1.1" LDFLAGS=" -L$PREFIX/lib/openssl <Your existing command>
```

Additionally CMake needs some extra command line flags too in order to build packages that need openssl 1.1 due to the fact that the header files and shared object libraries stored in non-standard paths.

```
cmake -DOPENSSL_INCLUDE_DIR=$PREFIX/include/openssl-1.1 -DOPENSSL_LIBRARIES=$PREFIX/lib/openssl-1.1 -DOPENSSL_CRYPTO_LIBRARY=$TERMUX_PREFIX/lib/openssl-1.1/libcrypto.so.1.1 -DOPENSSL_SSL_LIBRARY=$PREFIX/lib/openssl-1.1/libssl.so.1.1 <Other CMake options>
```

# Recovering from a broken environment

[Video instructions to recover from broken environment](https://user-images.githubusercontent.com/68849076/155350747-b13a7eeb-ae1d-4be2-bd92-5042b298218c.mp4)

> NOTE: First of all ensure that you are using latest version of Termux App

> ⚠️ ⚠️ ⚠️ WARNING ⚠️⚠️⚠️: These instructions will remove all your installed packages and also any package you installed from third party sources or any modifications you made yourself in $PREFIX. Nevertheless, your $HOME should remain untouched

1. Open a failsafe session and clear current `$PREFIX`

```
$ rm -rf $PREFIX
```

2. Re-open Termux. It should reinstall the bootstrap.

3. Switch to grimler.se mirrors

```
$ termux-change-repo
```

4. Update your packages (don't panic, grimler.se mirrors won't recieve updates until OpenSSL 3 rebuild is done)

```
$ pkg up
```

Enjoy!!

**Comment by thunder-coding** [comment]
Anyways @Grimler91, if you are ready to halt down updates for your repo when openssl 3 rebuild will be done, feel free to add the same note in the issue comment. The same way you did for NDK r23 update

**Comment by Grimler91** [comment]
@thunder-coding sure, can pause it for some time!

**Comment by truboxl** [comment]
Wonder if we need to provide a separate libcurl...

**Comment by thunder-coding** [comment]
> Wonder if we need to provide a separate libcurl...

Most probably not, any particular reason why this could be needed?

**Comment by thunder-coding** [comment]
> For example, the `php` package depends on both `libcurl` and `openssl`, and indeed the executable `bin/php` records both `libcurl.so` and `libssl.so.1.1` (and `libcrypto.so.1.1`) as `DT_NEEDED`. What happens if `php` itself links against `openssl-1.1` but its dependency `libcurl` links against `openssl-3.0`?

This will cause problems for php7, but not php 8. Probably we can make use of symbol versioning to solve the issue. https://github.com/archlinux/svntogit-packages/blob/packages/openssl-1.0/trunk/openssl-1.0-versioned-symbols.patch (Arch using symbol versioning for OpenSSL 1.0 legacy version)

**Comment by thunder-coding** [comment]
Anyways, I just found that both OpenSSL 1.1.1m and OpenSSL 3 use versioned symbols by default. So lubcurl should no longer be a problem

¯\_(ツ)_/¯

**Comment by librehat** [comment]
I wonder if we can package it as openssl3 and slowly move packages to depend on openssl3? Or package the previous one as openssl1.

Not sure if **ALL** packages can be compiled with OpenSSL 3 at this point.

**Comment by xtkoba** [comment]
@librehat The previous one will be renamed to `openssl-1.1`, according to the description.

**Comment by coolreader18** [comment]
Is there a CI link or something to follow when the rebuild will be done?

**Comment by Grimler91** [comment]
> Is there a CI link or something to follow when the rebuild will be done?

[The PR](https://github.com/termux/termux-packages/pull/8721) and accompanying CI builds will show when it is merged and builds completed

**Comment by thunder-coding** [comment]
All packages depending on OpenSSL have been rebuilt. In case anything is broken, consider opening a new issue.

**Comment by 2096779623** [comment]
MongoDB needs to be fixed:CANNOT LINK EXECUTABLE "/data/data/com.termux/files/usr/bin/mariadbd": library "libssl.so.3" not found: needed by main executable.

**Comment by albertz** [comment]
I just wanted to comment: I landed on this issue via a Google search. In my case ssh and other things were all broken and the error was with libgcrypt.so.3.

None of the suggestions from here were working.

But the real reason is that I installed Termux via the Google Play Store, and apparently that package is totally outdated and broken. Then I installed it via F-Droid and now all is fine.

**Comment by ElvenSpellmaker** [comment]
I had a problem with being able to use `pkg` as all mirrors failed and then the fail back curl fails due to ssl.

So I tried to follow the instructions and now regardless of the mirror I have:
```
[*] Grimler's mirrors selected                      Changing main repository
    Changing game repository                        Changing science repository
[*] Running apt update                          Hit:1 https://grimler.se/termux-packages-24 stable InRelease                                    Get:3 https://grimler.se/science-packages-24 science InRelease [9360 B]
Get:2 https://grimler.se/game-packages-24 games InRelease [9347 B]
Err:3 https://grimler.se/science-packages-24 science InRelease                                    The following signatures couldn't be verified because the public key is not available: NO_PUBKEY B0076E490B71616B                             Err:2 https://grimler.se/game-packages-24 games InRelease                                         The following signatures couldn't be verified because the public key is not available: NO_PUBKEY B0076E490B71616B                             Reading package lists... Done                   W: GPG error: https://grimler.se/science-packages-24 science InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY B0076E490B71616B       E: The repository 'https://grimler.se/science-packages-24 science InRelease' is not signed.     N: Metadata integrity can't be verified, repository is disabled now.                            N: Possible cause: repository is under maintenance or down (wrong sources.list URL?).
W: GPG error: https://grimler.se/game-packages-24 games InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY B0076E490B71616B
E: The repository 'https://grimler.se/game-packages-24 games InRelease' is not signed.
N: Metadata integrity can't be verified, repository is disabled now.
N: Possible cause: repository is under maintenance or down (wrong sources.list URL?).
```
